File: comments/src/main/java/net/particify/arsnova/comments/config/RabbitConfig.java
package net.particify.arsnova.comments.config;

import com.fasterxml.jackson.databind.ObjectMapper;
import org.springframework.amqp.core.Binding;
import org.springframework.amqp.core.BindingBuilder;
import org.springframework.amqp.core.Declarables;
import org.springframework.amqp.core.FanoutExchange;
import org.springframework.amqp.core.Queue;
import org.springframework.amqp.core.QueueBuilder;
import org.springframework.amqp.rabbit.connection.CachingConnectionFactory;
import org.springframework.amqp.rabbit.connection.ConnectionFactory;
import org.springframework.amqp.rabbit.core.RabbitTemplate;
import org.springframework.amqp.support.converter.Jackson2JsonMessageConverter;
import org.springframework.amqp.support.converter.MessageConverter;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.core.task.TaskExecutor;

@Configuration
public class RabbitConfig {
  public static final String BACKEND_COMMENT_FANOUT_NAME = "backend.event.comment.beforecreation";
  public static final String BACKEND_COMMENT_QUEUE_NAME = BACKEND_COMMENT_FANOUT_NAME + ".consumer.comment-service";
  public static final String BACKEND_ROOM_DUPLICATED_FANOUT_NAME = "backend.event.room.duplicated";
  public static final String BACKEND_ROOM_DUPLICATED_QUEUE_NAME = "backend.event.room.duplicated.consumer.comment-service";
  public static final String COMMENT_SERVICE_COMMENT_DELETE_FANOUT_NAME = "commentservice.event.comment.deleted";
  public static final String ROOM_CREATED_FANOUT_NAME = "backend.event.room.aftercreation";
  public static final String ROOM_CREATED_QUEUE_NAME = ROOM_CREATED_FANOUT_NAME + ".consumer.comment-service";
  public static final String ROOM_CREATED_DLQ_NAME = ROOM_CREATED_QUEUE_NAME + ".dlq";
  public static final String ROOM_DELETED_FANOUT_NAME = "backend.event.room.afterdeletion";
  public static final String ROOM_DELETED_QUEUE_NAME = ROOM_DELETED_FANOUT_NAME + ".consumer.comment-service";
  public static final String ROOM_DELETED_DLQ_NAME = ROOM_DELETED_QUEUE_NAME + ".dlq";

  @Value("${spring.rabbitmq.host}") private String rabbitmqHost;
  @Value("${spring.rabbitmq.port}") private int rabbitmqPort;
  @Value("${spring.rabbitmq.username}") private String rabbitmqUsername;
  @Value("${spring.rabbitmq.password}") private String rabbitmqPassword;
  @Value("${spring.rabbitmq.vhost}") private String rabbitmqVhost;

  @Bean
  public ConnectionFactory connectionFactory(
      @TaskExecutorConfig.RabbitConnectionExecutor TaskExecutor executor
  ) {
    final CachingConnectionFactory connectionFactory = new CachingConnectionFactory(rabbitmqHost, rabbitmqPort);
    connectionFactory.setUsername(rabbitmqUsername);
    connectionFactory.setPassword(rabbitmqPassword);
    connectionFactory.setVirtualHost(rabbitmqVhost);
    connectionFactory.setExecutor(executor);
    return connectionFactory;
  }

  @Bean
  public RabbitTemplate rabbitTemplate(ConnectionFactory connectionFactory, MessageConverter messageConverter) {
    final RabbitTemplate rabbitTemplate = new RabbitTemplate(connectionFactory);
    rabbitTemplate.setMessageConverter(messageConverter);
    return rabbitTemplate;
  }

  @Bean
  public MessageConverter jsonMessageConverter(final ObjectMapper objectMapper) {
    return new Jackson2JsonMessageConverter(objectMapper);
  }

  @Bean
  public Declarables rabbitDeclarables() {
    final FanoutExchange commentFanoutExchange = new FanoutExchange(BACKEND_COMMENT_FANOUT_NAME);
    final Queue commentQueue = new Queue(BACKEND_COMMENT_QUEUE_NAME, true, false, false);
    final Binding commentBinding = BindingBuilder.bind(commentQueue).to(commentFanoutExchange);
    final FanoutExchange roomDuplicatedFanoutExchange =  new FanoutExchange(BACKEND_ROOM_DUPLICATED_FANOUT_NAME);
    final Queue roomDuplicatedQueue =  new Queue(BACKEND_ROOM_DUPLICATED_QUEUE_NAME, true, false, false);
    final Binding roomDuplicatedBinding = BindingBuilder.bind(roomDuplicatedQueue).to(roomDuplicatedFanoutExchange);

    final FanoutExchange roomCreatedFanoutExchange = new FanoutExchange(ROOM_CREATED_FANOUT_NAME);
    final Queue roomCreatedDlq = new Queue(ROOM_CREATED_DLQ_NAME, true, false, false);
    final Queue roomCreatedQueue = QueueBuilder
      .durable(ROOM_CREATED_QUEUE_NAME)
      .deadLetterExchange("")
      .deadLetterRoutingKey(ROOM_CREATED_DLQ_NAME)
      .build();
    final Binding roomCreatedBinding = BindingBuilder.bind(roomCreatedQueue).to(roomCreatedFanoutExchange);

    final FanoutExchange roomDeletedFanoutExchange = new FanoutExchange(ROOM_DELETED_FANOUT_NAME);
    final Queue roomDeletedDlq = new Queue(ROOM_DELETED_DLQ_NAME, true, false, false);
    final Queue roomDeletedQueue = QueueBuilder
        .durable(ROOM_DELETED_QUEUE_NAME)
        .deadLetterExchange("")
        .deadLetterRoutingKey(ROOM_DELETED_DLQ_NAME)
        .build();
    final Binding roomDeletedBinding = BindingBuilder.bind(roomDeletedQueue).to(roomDeletedFanoutExchange);

    final FanoutExchange deleteFanoutExchange = new FanoutExchange(COMMENT_SERVICE_COMMENT_DELETE_FANOUT_NAME);

    return new Declarables(
        commentFanoutExchange,
        commentQueue,
        commentBinding,
        deleteFanoutExchange,
        roomCreatedFanoutExchange,
        roomCreatedDlq,
        roomCreatedQueue,
        roomCreatedBinding,
        roomDeletedFanoutExchange,
        roomDeletedDlq,
        roomDeletedQueue,
        roomDeletedBinding,
        roomDuplicatedFanoutExchange,
        roomDuplicatedQueue,
        roomDuplicatedBinding
    );
  }
}

File: authz/src/main/kotlin/net/particify/arsnova/authz/config/RabbitConfig.kt
package net.particify.arsnova.authz.config

import net.particify.arsnova.authz.model.event.RoomAccessSyncEvent
import net.particify.arsnova.authz.model.event.RoomAccessSyncRequest
import org.springframework.amqp.core.BindingBuilder.bind
import org.springframework.amqp.core.Declarables
import org.springframework.amqp.core.DirectExchange
import org.springframework.amqp.core.FanoutExchange
import org.springframework.amqp.core.Queue
import org.springframework.amqp.rabbit.connection.CachingConnectionFactory
import org.springframework.amqp.rabbit.connection.ConnectionFactory
import org.springframework.amqp.rabbit.core.RabbitAdmin
import org.springframework.amqp.rabbit.core.RabbitTemplate
import org.springframework.amqp.support.converter.ClassMapper
import org.springframework.amqp.support.converter.DefaultClassMapper
import org.springframework.amqp.support.converter.JacksonJsonMessageConverter
import org.springframework.amqp.support.converter.MessageConverter
import org.springframework.boot.context.properties.EnableConfigurationProperties
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.core.task.TaskExecutor
import tools.jackson.databind.json.JsonMapper

@Configuration
@EnableConfigurationProperties(AuthServiceProperties::class)
class RabbitConfig(
  private var authServiceProperties: AuthServiceProperties,
) {
  companion object {
    const val ROOM_ACCESS_SYNC_RESPONSE_QUEUE_NAME: String = "backend.event.room.access.sync.response"
    const val ROOM_ACCESS_SYNC_REQUEST_QUEUE_NAME: String = "backend.event.room.access.sync.request"
    const val ROOM_CREATED_EXCHANGE_NAME: String = "backend.event.room.aftercreation"
    const val ROOM_DELETED_EXCHANGE_NAME: String = "backend.event.room.afterdeletion"
    const val ROOM_CREATED_QUEUE_NAME: String = "backend.event.room.aftercreation.consumer.auth-service"
    const val ROOM_DELETED_QUEUE_NAME: String = "backend.event.room.afterdeletion.consumer.auth-service"
    const val PARTICIPANT_ACCESS_MIGRATION_QUEUE_NAME: String = "backend.event.migration.access.participant"
  }

  @Bean
  fun connectionFactory(
    @TaskExecutorConfig.RabbitConnectionExecutor executor: TaskExecutor,
  ): ConnectionFactory {
    val connectionFactory =
      CachingConnectionFactory(
        authServiceProperties.rabbitmq.host,
        authServiceProperties.rabbitmq.port,
      )
    connectionFactory.username = authServiceProperties.rabbitmq.username
    connectionFactory.setPassword(authServiceProperties.rabbitmq.password)
    connectionFactory.virtualHost = authServiceProperties.rabbitmq.virtualHost
    connectionFactory.setExecutor(executor)
    return connectionFactory
  }

  @Bean
  fun rabbitTemplate(
    connectionFactory: ConnectionFactory,
    messsageConverter: MessageConverter,
  ): RabbitTemplate {
    val rabbitTemplate = RabbitTemplate(connectionFactory)
    rabbitTemplate.messageConverter = messsageConverter
    return rabbitTemplate
  }

  @Bean
  fun rabbitAdmin(connectionFactory: ConnectionFactory): RabbitAdmin = RabbitAdmin(connectionFactory)

  @Bean
  fun declarables(): Declarables {
    val roomCreatedFanoutExchange = FanoutExchange(ROOM_CREATED_EXCHANGE_NAME)
    val roomDeletedFanoutExchange = FanoutExchange(ROOM_DELETED_EXCHANGE_NAME)
    val participantAccessMigrationDlq = Queue("$PARTICIPANT_ACCESS_MIGRATION_QUEUE_NAME.dlq")
    val participantAccessMigrationQueue =
      Queue(
        PARTICIPANT_ACCESS_MIGRATION_QUEUE_NAME,
        true,
        false,
        false,
        mapOf(
          "x-dead-letter-exchange" to "",
          "x-dead-letter-routing-key" to "$PARTICIPANT_ACCESS_MIGRATION_QUEUE_NAME.dlq",
        ),
      )
    val roomCreatedQueue =
      Queue(
        ROOM_CREATED_QUEUE_NAME,
        true,
        true,
        false,
        mapOf(
          "x-dead-letter-exchange" to "",
          "x-dead-letter-routing-key" to "$ROOM_CREATED_QUEUE_NAME.dlq",
        ),
      )
    val roomDeletedQueue =
      Queue(
        ROOM_DELETED_QUEUE_NAME,
        true,
        true,
        false,
        mapOf(
          "x-dead-letter-exchange" to "",
          "x-dead-letter-routing-key" to "$ROOM_DELETED_QUEUE_NAME.dlq",
        ),
      )
    val roomCreatedDlq = Queue("$ROOM_CREATED_QUEUE_NAME.dlq")
    val roomDeletedDlq = Queue("$ROOM_DELETED_QUEUE_NAME.dlq")
    val roomAccessSyncResponseQueue =
      Queue(
        ROOM_ACCESS_SYNC_RESPONSE_QUEUE_NAME,
        true,
        false,
        false,
        mapOf(
          "x-dead-letter-exchange" to "",
          "x-dead-letter-routing-key" to "$ROOM_ACCESS_SYNC_RESPONSE_QUEUE_NAME.dlq",
        ),
      )
    val roomAccessSyncResponseDlq = Queue("$ROOM_ACCESS_SYNC_RESPONSE_QUEUE_NAME.dlq")
    val roomAccessSyncRequestExchange = DirectExchange(ROOM_ACCESS_SYNC_REQUEST_QUEUE_NAME)

    return Declarables(
      listOf(
        roomCreatedFanoutExchange,
        roomDeletedFanoutExchange,
        participantAccessMigrationQueue,
        participantAccessMigrationDlq,
        roomCreatedQueue,
        roomCreatedDlq,
        roomDeletedQueue,
        roomDeletedDlq,
        bind(roomCreatedQueue).to(roomCreatedFanoutExchange),
        bind(roomDeletedQueue).to(roomDeletedFanoutExchange),
        roomAccessSyncResponseQueue,
        roomAccessSyncResponseDlq,
        roomAccessSyncRequestExchange,
      ),
    )
  }

  @Bean
  fun jsonMessageConverter(jsonMapper: JsonMapper): MessageConverter = JacksonJsonMessageConverter(jsonMapper)

  @Bean
  fun classMapper(): ClassMapper {
    val classMapper = DefaultClassMapper()
    val idClassMapping: MutableMap<String, Class<*>> = HashMap()
    // Because of not having shared models we need this mapping to backend models for jackson
    idClassMapping["net.particify.arsnova.core.event.RoomAccessSyncRequest"] = RoomAccessSyncRequest::class.java
    idClassMapping["net.particify.arsnova.core.event.RoomAccessSyncEvent"] = RoomAccessSyncEvent::class.java
    classMapper.setIdClassMapping(idClassMapping)
    return classMapper
  }
}
